File: solutions/cpp/Matrix_search.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <vector>
#include <limits>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::max;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

// @include
bool MatrixSearch(const vector<vector<int>>& A, int x) {
  int row = 0, col = A[0].size() - 1;  // Start from the top-right corner.
  // Keeps searching while there are unclassified rows and columns.
  while (row < A.size() && col >= 0) {
    if (A[row][col] == x) {
      return true;
    } else if (A[row][col] < x) {
      ++row;  // Eliminate this row.
    } else {  // A[row][col] > x.
      --col;  // Eliminate this column.
    }
  }
  return false;
}
// @exclude

static void SimpleTest() {
  vector<int> A0 = {1};
  vector<vector<int>> A; 
  A[0] = A0;
  assert(!MatrixSearch(A, 0));
  assert(MatrixSearch(A, 1));

  A0 = {1,5};
  vector<int> A1 = {2,6};
  A[0] = A0;
  A[1] = A1;
  assert(!MatrixSearch(A, 0));
  assert(MatrixSearch(A, 1));
  assert(MatrixSearch(A, 2));
  assert(MatrixSearch(A, 5));
  assert(MatrixSearch(A, 6));
  assert(!MatrixSearch(A, 3));
  assert(!MatrixSearch(A, INT_MAX));

  A0[0] = 2;
  assert(!MatrixSearch(A, 1));
  assert(MatrixSearch(A, 2));

  A0 = {1,5,7};
  A1 = {3,10,100};
  vector<int> A2 = {3,12,INT_MAX};
  A[0] = A0;
  A[1] = A1;
  A[2] = A2;
  assert(MatrixSearch(A, 1));
  assert(!MatrixSearch(A, 2));
  assert(!MatrixSearch(A, 4));
  assert(MatrixSearch(A, 3));
  assert(MatrixSearch(A, 10));
  assert(MatrixSearch(A, INT_MAX));
  assert(!MatrixSearch(A, INT_MAX-1));
  assert(MatrixSearch(A, 12));
}



// O(n^2) solution for verifying answer.
bool BruteForceSearch(const vector<vector<int>>& A, int x) {
  for (int i = 0; i < A.size(); ++i) {
    for (int j = 0; j < A[i].size(); ++j) {
      if (A[i][j] == x) {
        return true;
      }
    }
  }
  return false;
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 10000; ++times) {
    int n, m;
    if (argc == 3) {
      n = atoi(argv[1]), m = atoi(argv[2]);
    } else {
      uniform_int_distribution<int> dis(1, 100);
      n = dis(gen), m = dis(gen);
    }
    vector<vector<int>> A(n, vector<int>(m));
    uniform_int_distribution<int> dis(0, 99);
    A[0][0] = dis(gen);
    for (int i = 0; i < n; ++i) {
      for (int j = 0; j < m; ++j) {
        int up = (i == 0) ? 0 : A[i - 1][j];
        int left = (j == 0) ? 0 : A[i][j - 1];
        uniform_int_distribution<int> shift_dis(1, 20);
        A[i][j] = max(up, left) + shift_dis(gen);
      }
    }
    uniform_int_distribution<int> x_dis(0, 999);
    int x = x_dis(gen);
    assert(BruteForceSearch(A, x) == MatrixSearch(A, x));
  }
  return 0;
}

File: solutions/python/matrix_search.py
import sys
import random


# @include
def matrix_search(A, x):
    row, col = 0, len(A[0]) - 1  # Start from the top-right corner.
    # Keeps searching while there are unclassified rows and columns.
    while row < len(A) and col >= 0:
        if A[row][col] == x:
            return True
        elif A[row][col] < x:
            row += 1  # Eliminate this row.
        else:  # A[row][col] > x.
            col -= 1  # Eliminate this column.
    return False
# @exclude


def simple_test():
    A = [[1]]
    assert not matrix_search(A, 0)
    assert matrix_search(A, 1)

    A = [[1, 5], [2, 6]]
    assert not matrix_search(A, 0)
    assert matrix_search(A, 1)
    assert matrix_search(A, 2)
    assert matrix_search(A, 5)
    assert matrix_search(A, 6)
    assert not matrix_search(A, 3)
    assert not matrix_search(A, float('inf'))

    A = [[2, 5], [2, 6]]
    assert not matrix_search(A, 1)
    assert matrix_search(A, 2)

    A = [[1, 5, 7], [3, 10, 100], [3, 12, float('inf')]]
    assert matrix_search(A, 1)
    assert not matrix_search(A, 2)
    assert not matrix_search(A, 4)
    assert matrix_search(A, 3)
    assert matrix_search(A, 10)
    assert matrix_search(A, float('inf'))
    assert matrix_search(A, 12)


# O(n^2) solution for verifying answer.
def brute_force_search(A, x):
    for row in A:
        for cell in row:
            if cell == x:
                return True
    return False


def main():
    simple_test()
    for _ in range(10000):
        if len(sys.argv) == 3:
            n = int(sys.argv[1])
            m = int(sys.argv[2])
        else:
            n = random.randint(1, 100)
            m = random.randint(1, 100)
        A = [[0] * m for i in range(n)]
        A[0][0] = random.randrange(100)
        for i in range(n):
            for j in range(m):
                up = 0 if i == 0 else A[i - 1][j]
                left = 0 if j == 0 else A[i][j - 1]
                A[i][j] = max(up, left) + random.randint(1, 20)
        x = random.randrange(100)
        assert brute_force_search(A, x) == matrix_search(A, x)


if __name__ == '__main__':
    main()
